include/ISceneNode.h

```diff
@@ -276,30 +277,31 @@ namespace scene
 				child->grab();
 				child->remove(); // remove from old parent
 				Children.push_back(child);
+				// Note: This iterator is not invalidated until we erase it.
+				child->ThisIterator = Children.end();
+				--(*child->ThisIterator);
 				child->Parent = this;
 			}
 		}
 
 
 		//! Removes a child from this scene node.
-		/** If found in the children list, the child pointer is also
-		dropped and might be deleted if no other grab exists.
+		/**
 		\param child A pointer to the child which shall be removed.
 		\return True if the child was removed, and false if not,
-		e.g. because it couldn't be found in the children list. */
+		e.g. because it belongs to a different parent or no parent. */
 		virtual bool removeChild(ISceneNode* child)
 		{
-			ISceneNodeList::iterator it = Children.begin();
-			for (; it != Children.end(); ++it)
-				if ((*it) == child)
-				{
-					(*it)->Parent = 0;
-					(*it)->drop();
-					Children.erase(it);
-					return true;
-				}
+			if (child->Parent != this)
+				return false;
 
-			return false;
+			// The iterator must be set since the parent is not null.
+			auto it = *child->ThisIterator;
```

Comment:
prefer `.value()` here since that checks

Reply:
I would use that, but don't we still compile Irrlicht with exceptions disabled? Will it `abort` if the check fails?

Reply:
:shrug:
you could also put `_IRR_DEBUG_BREAK_IF(!child->ThisIterator.has_value())` here (which we'll convert to an assert one day)